tawazi/node/node.py

```diff
@@ -77,71 +78,89 @@ class ExecNode:
+        priority (int): priority compared to other ExecNodes; the higher the number the higher the priority.
+        is_sequential (bool): whether to execute this ExecNode in sequential order with respect to others.
+            When this ExecNode must be executed, all other nodes are waited to finish before starting execution.
+            Defaults to False.
+        debug (bool): Make this ExecNode a debug Node. Defaults to False.
+        tag (TagOrTags): Attach a Tag or Tags to this ExecNode. Defaults to None.
+        setup (bool): Make this ExecNode a setup Node. Defaults to False.
+        unpack_to (Optional[int]): if not None, this ExecNode's execution must return unpacked results corresponding
             to the given value
-            resource (str): the resource to use to execute this ExecNode. Defaults to "thread".
+        resource (str): the resource to use to execute this ExecNode. Defaults to "thread".
 
-        Raises:
-            ValueError: if setup and debug are both True.
-        """
-        # 1. assign attributes
-        self._id = id_
-        self.exec_function = exec_function
-        self.priority = priority
-        self.is_sequential = is_sequential
-        self.debug = debug
-        self.tag = tag
-        self.setup = setup
-        self.unpack_to = unpack_to
-        self.active = True
-        self.resource = resource
-
-        self.args: List[UsageExecNode] = args or []
-        self.kwargs: Dict[Identifier, UsageExecNode] = kwargs or {}
-
-        # 3. Assign the name
-        # This can be used in the future but is not particularly useful at the moment
-        self.__name__ = self.exec_function.__name__ if not isinstance(id_, str) else id_
-
-        # 4. Assign a default NoVal to the result of the execution of this ExecNode,
-        #  when this ExecNode will be executed, self.result will be overridden
-        # It would be amazing if we can remove self.result and make ExecNode immutable
-        self.result: Union[NoValType, Any] = NoVal
-        """Internal attribute to store the result of the execution of this ExecNode (Might change!)."""
-        # even though setting result to NoVal is not necessary... it clarifies debugging
+    Raises:
+        ValueError: if setup and debug are both True.
+    """
 
-        self.profile = Profile(cfg.TAWAZI_PROFILE_ALL_NODES)
+    # 1. assign attributes
+    id_: Identifier
+    exec_function: Callable[..., Any] = field(default_factory=lambda: lambda *args, **kwargs: None)
+    priority: int = 0
+    is_sequential: bool = cfg.TAWAZI_IS_SEQUENTIAL
+    debug: bool = False
+    tag: Optional[TagOrTags] = None
+    setup: bool = False
+    unpack_to: Optional[int] = None
+    resource: Resource = cfg.TAWAZI_DEFAULT_RESOURCE
+
+    args: List["UsageExecNode"] = field(default_factory=list)  # args or []
+    kwargs: Dict[Identifier, "UsageExecNode"] = field(default_factory=dict)  # kwargs or {}
+
+    # TODO: fix _active behavior!
```

Comment:
what is the problem in this TODO ?

Reply:
currently whne an ExecNode is not active, all its successors are removed. This makes it unusable in real world scenarios